sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Util.scala

```diff
@@ -597,11 +597,31 @@ private[sql] object CatalogV2Util {
       // Note: the back-fill here is a logical concept. The data source can keep the existing
       //       data unchanged and let the data reader to return "exist default" for missing
       //       columns.
-      val existingDefault = Literal(default.getValue.value(), default.getValue.dataType()).sql
-      f.withExistenceDefaultValue(existingDefault).withCurrentDefaultValue(default.getSql)
+      val existsDefault = extractExistsDefault(default)
+      val (sql, expr) = extractCurrentDefault(default)
+      val newMetadata = new MetadataBuilder()
+        .withMetadata(f.metadata)
+        .putString(EXISTS_DEFAULT_COLUMN_METADATA_KEY, existsDefault)
+        .putExpression(CURRENT_DEFAULT_COLUMN_METADATA_KEY, sql, expr)
+        .build()
+      f.copy(metadata = newMetadata)
     }.getOrElse(f)
   }
 
+  private def extractExistsDefault(default: ColumnDefaultValue): String = {
+    Literal(default.getValue.value(), default.getValue.dataType()).sql
+  }
+
+  private def extractCurrentDefault(default: ColumnDefaultValue): (String, Option[Expression]) = {
+    val expr = Option(default.getExpression).flatMap(V2ExpressionUtils.toCatalyst)
+    val sql = Option(default.getSql).orElse(expr.map(_.sql)).getOrElse {
```

Comment:
[doubt] my understanding was `.sql` is not reliable (based on discussion [here](https://github.com/apache/spark/pull/50792#discussion_r2078776399)), wondering if this could lead to users using `getMap` or map.get(key), directly and extracting the SQL from the map, skip actually checking if there is an expression for it and one should use that instead ? essentially if there i an entry in the runtimeMap should we let the map.get fail ?

Reply:
I don't think we mean these properties to be accessible by users. We do that for built-in tables but there are proper DSv2 APIs for this like `ColumnDefaultValue` where the expression always takes precedence.  SQL is informational in this case.